roadmap/implementers-guide/src/types/approval.md

```diff
@@ -0,0 +1,128 @@
+# Approval Types
+
+## ApprovalId
+
+The public key of a keypair used by a validator for approval voting.
+
+## AssignmentId
+
+The private key of a keypair used by a validator for approval voting.
+
+## AssignmentCert
+
+An `AssignmentCert`, short for Assignment Certificate, is a piece of data provided by a validator to prove that they have been selected to perform secondary approval checks on an included candidate.
+
+These certificates can be checked in the context of a specific block, candidate, and validator assignment VRF key. The block state will also provide further context about the availability core states at that block.
+
+```rust
+enum AssignmentCertKind {
+    RelayVRFModulo {
+        relay_vrf: (VRFInOut, VRFProof),
+        sample: u32,
+    },
+    RelayVRFDelay {
+        relay_vrf: (VRFInOut, VRFProof),
+        core_index: CoreIndex,
+    }
+}
+
+struct AssignmentCert {
+    // The criterion which is claimed to be met by this cert.
+    kind: AssignmentCertKind,
+    // The VRF showing the criterion is met.
+    vrf: VRFInOut,
+}
```

Comment:
You've the `vrf_inout` living in two places here. 

I'd normally have a `vrf_preout` and `vrf_proof` together ala https://github.com/paritytech/polkadot/pull/1558/files#diff-16fadbdcb87a7ecb8febad865d8ca383f4c913f4bf272307bdd93b7a0a3ab26cR233 but a vrf_inout would live in the type either before signing or after verification ala https://github.com/paritytech/polkadot/pull/1558/files#diff-16fadbdcb87a7ecb8febad865d8ca383f4c913f4bf272307bdd93b7a0a3ab26cR161

I jumped through polymorphism hoops to unify the notion of an assignment before signing or after verification with the idea that I'd unify some data structures and make fewer mistakes.  It did not go as well as I'd envisioned though, so I ended up with the expected three final types largely separate:
- our assignment with `vrf_inout` computed to find delay tranche, but not yet signed because maybe never used
- actually signed assignments, and
- verified assignments by anyone, ala your `TrancheEntry`.

Reply:
Can you clarify what you would expect to have for the _unverified_ type, which is this `AssignmentCert`?

Reply:
An unverified would contain a `VRFPreOut` and `VRFProof` because although you could compute the `VRFInOut` earlier it'll make you code robust if you never compute the `VRFInOut` until you call `vrf_verify`/`dleq_verifiy`. 

An assignment of our own that is not yet announced would contain a `VRFInOut` for which we did not yet bother to compute the `VRFProof`.  I'd originally meant to unify this into our own unannounced being handled in parallel with validated ones from others, but that did not go perfectly.

I'm unsure how you want to buffer the unsent gossip messages.  I ended up holding them in the same data structures as things containing both a `VRFProof` and a `VRFInOut`, from which one can easily compute the `VRFPreOut` required in gossip.  That kinda centralized all the timing/tranche accounting, but I suspect you guys have separate modules for such things.

Reply:
Also, I'd envisioned using the VRF as the gossip message signature, which saves doing another signature, but complicates the layers since we'd verify the VRF even to know if the message makes sense.  We could avoid this verification triggering any chain historical lookups by incluidng the `VRFPreOut` along with the message, and then validating that later.  It's too much code subtelty to do this maybe..